hubspot/client.go

```diff
@@ -0,0 +1,231 @@
+		Timeout: 30 * time.Second,
+		Transport: &http.Transport{
+			Dial: (&net.Dialer{
+				Timeout: 10 * time.Second,
+			}).Dial,
+		},
+	}
+
+	a.chStopped = make(chan struct{})
+
+	a.wgSenders.Add(1)
+	go a.fetchEvents()
+
+	go func() {
+		a.wgSenders.Wait()
+		close(a.chStopped)
+	}()
+
+	return a, a.chStopped, nil
+}
+
+func (a *HubSpotAdapter) Close() error {
+	a.conf.ClientOptions.DebugLog("closing")
+	a.doStop.Set()
+	a.wgSenders.Wait()
+	err1 := a.uspClient.Drain(1 * time.Minute)
+	_, err2 := a.uspClient.Close()
+	a.httpClient.CloseIdleConnections()
+
+	if err1 != nil {
+		return err1
+	}
+
+	return err2
+}
+
+func (a *HubSpotAdapter) fetchEvents() {
+	defer a.wgSenders.Done()
+	defer a.conf.ClientOptions.DebugLog(fmt.Sprintf("fetching of %s events exiting", logsURL))
+
+	adapterStart := time.Now()
+	for !a.doStop.WaitFor(30 * time.Second) {
+		// The makeOneRequest function handles error
+		// handling and fatal error handling.
+		items := a.makeOneRequest(adapterStart)
+		if items == nil {
+			continue
+		}
+
+		for _, item := range items {
+			msg := &protocol.DataMessage{
+				JsonPayload: item,
+				TimestampMs: uint64(time.Now().UnixNano() / int64(time.Millisecond)),
+			}
+			if err := a.uspClient.Ship(msg, 10*time.Second); err != nil {
+				if err == uspclient.ErrorBufferFull {
+					a.conf.ClientOptions.OnWarning("stream falling behind")
+					err = a.uspClient.Ship(msg, 1*time.Hour)
+				}
+				if err == nil {
+					continue
+				}
+				a.conf.ClientOptions.OnError(fmt.Errorf("Ship(): %v", err))
+				a.doStop.Set()
+				return
+			}
+		}
+	}
+}
+
+func (a *HubSpotAdapter) makeOneRequest(notBefore time.Time) []utils.Dict {
+	var allItems []utils.Dict
+	currentTime := time.Now()
+	var start string
+
+	if t := currentTime.Add(-overlapPeriod); t.Before(notBefore) {
+		start = notBefore.UTC().Format(time.RFC3339)
+	} else {
+		start = currentTime.Add(-overlapPeriod).UTC().Format(time.RFC3339)
+	}
+	until := currentTime.UTC().Format(time.RFC3339)
+
+	for {
+		// Prepare the request.
+		req, err := http.NewRequest("GET", fmt.Sprintf("%s?occurredAfter=%s&occurredBefore=%s", logsURL, start, until), nil)
+		a.conf.ClientOptions.DebugLog(fmt.Sprintf("requesting from %s starting at %s until %s", logsURL, start, until))
+		if err != nil {
+			a.doStop.Set()
+			return nil
+		}
+		req.Header.Set("Content-Type", "application/json")
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", a.conf.AccessToken))
+
+		// Issue the request.
+		resp, err := a.httpClient.Do(req)
+		if err != nil {
+			a.conf.ClientOptions.OnError(fmt.Errorf("http.Client.Do(): %v", err))
+			return nil
+		}
+		defer resp.Body.Close()
+
+		// Evaluate if success.
+		if resp.StatusCode != http.StatusOK {
+			body, _ := ioutil.ReadAll(resp.Body)
+			a.conf.ClientOptions.OnError(fmt.Errorf("hubspot api non-200: %s\nREQUEST: %s\nRESPONSE: %s", resp.Status, string(body), string(body)))
+			return nil
+		}
+
+		body, _ := ioutil.ReadAll(resp.Body)
```

Comment:
we should try to get the error here and handle it
```suggestion
		body, err := ioutil.ReadAll(resp.Body)
```